#### Continuation from previous notebook

Recall from second notebook: 'Find a base and add on new features'. 

Created a small set of new features in the third notebook. 

In the final (fourth) notebook, run models on the 'final' dataset.

Run models: 
   1. Random Forest 
   2. Logistic Regression  
   3. SGD
   4. GBC
   5. AdaBoost Decision Tree.
   6. Ensemble of ensembles (voting=soft).
   7. Ensemble of 'best' (voting=soft).

In [3]:
## IMPORTS ##

# numpy and pandas for data manipulation
import numpy as np
import pandas as pd 

# sklearn preprocessing for dealing with categorical variables
from sklearn.preprocessing import LabelEncoder

# File system manangement
import os

# Suppress warnings 
import warnings
warnings.filterwarnings('ignore')

# matplotlib for plotting
import matplotlib.pyplot as plt

# garbage collector
import gc

In [4]:
# Load data fcn
def load_credit_data(data_path):
    csv_path = os.path.join("data", data_path)
    return pd.read_csv(csv_path)

In [5]:
# Load training data
# training_df = load_credit_data ("training_top.csv")
training_df = load_credit_data ("training_top_id.csv")
print (training_df.shape)

(307511, 268)


In [6]:
# Load features for training
features_training_df = load_credit_data ("training_new_features_v1.csv")
print (features_training_df.shape)
features_training_df.head()

(307511, 92)


SK_ID_CURR  APP_CREDIT_TO_ANNUITY_RATIO  APP_CREDIT_TO_GOODS_RATIO  \
0      100002                    16.461104                   1.158397   
1      100003                    36.234085                   1.145199   
2      100004                    20.000000                   1.000000   
3      100006                    10.532818                   1.052803   
4      100007                    23.461618                   1.000000   

   APP_INCOME_PER_CHLD  APP_CREDIT_TO_INCOME_RATIO  \
0             202500.0                    2.007889   
1             270000.0                    4.790750   
2              67500.0                    2.000000   
3             135000.0                    2.316167   
4             121500.0                    4.222222   

   APP_ANNUITY_TO_INCOME_RATIO  APP_SOURCES_PROD  APP_EXT_SOURCES_MEAN  \
0                     0.121977          0.003043              0.161787   
1                     0.132216          0.119932              0.466757   
2                     0.099999          0.119932              0.642739   
3                     0.219898          0.119932              0.650442   
4                     0.179961          0.119932              0.322738   

   APP_EXT_SCORES_STD  APP_CAR_TO_BIRTH_RATIO  \
0            0.092026               -0.000627   
1            0.219895               -0.000627   
2            0.122792               -0.001365   
3            0.136021               -0.000627   
4            0.136021               -0.000627   

                        ...                         \
0                       ...                          
1                       ...                          
2                       ...                          
3                       ...                          
4                       ...                          

   INSTALL_DAYS_ENTRY_PAYMENT_MEDIAN  INSTALL_DAYS_INSTALMENT_MAX  \
0                             -312.0                        -25.0   
1                             -806.0                       -536.0   
2                             -763.0                       -724.0   
3                             -211.0                        -11.0   
4                             -852.5                        -14.0   

   INSTALL_DAYS_INSTALMENT_MIN  INSTALL_DAYS_INSTALMENT_MEDIAN  \
0                       -565.0                          -295.0   
1                      -2310.0                          -797.0   
2                       -784.0                          -754.0   
3                       -545.0                          -206.0   
4                      -2326.0                          -851.0   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MAX  \
0                                       1.0   
1                                       1.0   
2                                       1.0   
3                                       1.0   
4                                       1.0   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MIN  \
0                                   1.00000   
1                                   1.00000   
2                                   1.00000   
3                                   1.00000   
4                                   0.00005   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MEAN  \
0                                   1.000000   
1                                   1.000000   
2                                   1.000000   
3                                   1.000000   
4                                   0.954545   

   INSTALL_INSTPAY_DIFF_DAYS_INSTALL_DAYS_PAY_MAX  \
0                                            31.0   
1                                            14.0   
2                                            11.0   
3                                            77.0   
4                                            31.0   

   INSTALL_INSTPAY_DIFF_DAYS_INSTALL_DAYS_PAY_MIN  \
0                                            12.0   
1                                             1.0   
2                            

In [7]:
# Merge for full set
# training_df = pd.concat ([training_df, features_training_df], axis=1)
training_df_merged = training_df.merge(right=features_training_df, on = 'SK_ID_CURR', how = 'left')

print (training_df_merged.shape)
training_df_merged.head()

(307511, 359)


SK_ID_CURR  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0      100002             0          202500.0    406597.5      24700.5   
1      100003             0          270000.0   1293502.5      35698.5   
2      100004             0           67500.0    135000.0       6750.0   
3      100006             0          135000.0    312682.5      29686.5   
4      100007             0          121500.0    513000.0      21865.5   

   REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  \
0                    0.018801       -9461           -637            -3648.0   
1                    0.003541      -16765          -1188            -1186.0   
2                    0.010032      -19046           -225            -4260.0   
3                    0.008019      -19005          -3039            -9833.0   
4                    0.028663      -19932          -3038            -4311.0   

   DAYS_ID_PUBLISH                       ...                         \
0            -2120                       ...                          
1             -291                       ...                          
2            -2531                       ...                          
3            -2437                       ...                          
4            -3458                       ...                          

   INSTALL_DAYS_ENTRY_PAYMENT_MEDIAN  INSTALL_DAYS_INSTALMENT_MAX  \
0                             -312.0                        -25.0   
1                             -806.0                       -536.0   
2                             -763.0                       -724.0   
3                             -211.0                        -11.0   
4                             -852.5                        -14.0   

   INSTALL_DAYS_INSTALMENT_MIN  INSTALL_DAYS_INSTALMENT_MEDIAN  \
0                       -565.0                          -295.0   
1                      -2310.0                          -797.0   
2                       -784.0                          -754.0   
3                       -545.0                          -206.0   
4                      -2326.0                          -851.0   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MAX  \
0                                       1.0   
1                                       1.0   
2                                       1.0   
3                                       1.0   
4                                       1.0   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MIN  \
0                                   1.00000   
1                                   1.00000   
2                                   1.00000   
3                                   1.00000   
4                                   0.00005   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MEAN  \
0                                   1.000000   
1                                   1.000000   
2                                   1.000000   
3                                   1.000000   
4                                   0.954545   

   INSTALL_INSTPAY_DIFF_DAYS_INSTALL_DAYS_PAY_MAX  \
0                                            31.0   
1                                            14.0   
2                                            11.0   
3                                            77.0   
4                                            31.0   

   INSTALL_INSTPAY_DIFF_DAYS_INSTALL_DAYS_PAY_MIN  \
0                                            12.0   
1                                             1.0   
2                                             3.0   
3                                             1.0   
4                                           -12.0   

   INSTALL_INSTPAY_DIFF_DAYS_INSTALL_DAYS_PAY_MEAN  
0                                        20.421053  
1                                         7.160000  
2                                         7.666667  
3                                        19.375000  
4                                         3.636364  

[5 rows x 359 columns]

In [8]:
# Load testing data
testing_df = load_credit_data ("testing_top_id.csv")
print (testing_df.shape)

(48744, 268)


In [9]:
# Load features for testing
features_testing_df = load_credit_data ("testing_new_features_v1.csv")
print (features_testing_df.shape)
features_testing_df.head()

(48744, 92)


SK_ID_CURR  APP_CREDIT_TO_ANNUITY_RATIO  APP_CREDIT_TO_GOODS_RATIO  \
0      100001                    27.664697                     1.2640   
1      100005                    12.824870                     1.2376   
2      100013                     9.505482                     1.0528   
3      100028                    32.130726                     1.0000   
4      100038                    19.506034                     1.0000   

   APP_INCOME_PER_CHLD  APP_CREDIT_TO_INCOME_RATIO  \
0             135000.0                    4.213333   
1              99000.0                    2.250182   
2             202500.0                    3.275378   
3             105000.0                    5.000000   
4              90000.0                    3.475000   

   APP_ANNUITY_TO_INCOME_RATIO  APP_SOURCES_PROD  APP_EXT_SOURCES_MEAN  \
0                     0.152299          0.094803              0.567263   
1                     0.175453          0.071345              0.429869   
2                     0.344576          0.119686              0.655389   
3                     0.155614          0.164177              0.549372   
4                     0.178149          0.119686              0.313916   

   APP_EXT_SCORES_STD  APP_CAR_TO_BIRTH_RATIO  \
0            0.353601               -0.000624   
1            0.136694               -0.000624   
2            0.062788               -0.000250   
3            0.055432               -0.000624   
4            0.158068               -0.001227   

                        ...                         \
0                       ...                          
1                       ...                          
2                       ...                          
3                       ...                          
4                       ...                          

   INSTALL_DAYS_ENTRY_PAYMENT_MEDIAN  INSTALL_DAYS_INSTALMENT_MAX  \
0                            -1715.0                      -1619.0   
1                             -585.0                       -466.0   
2                            -1383.0                        -14.0   
3                             -812.0                        -27.0   
4                             -634.0                       -457.0   

   INSTALL_DAYS_INSTALMENT_MIN  INSTALL_DAYS_INSTALMENT_MEDIAN  \
0                      -2916.0                         -1709.0   
1                       -706.0                          -586.0   
2                      -2705.0                         -1383.0   
3                      -1773.0                          -812.0   
4                       -787.0                          -622.0   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MAX  \
0                                       1.0   
1                                       1.0   
2                                       1.0   
3                                       1.0   
4                                       1.0   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MIN  \
0                                  1.000000   
1                                  1.000000   
2                                  0.000266   
3                                  0.030496   
4                                  1.000000   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MEAN  \
0                                   1.000000   
1                                   1.000000   
2                                   0.935484   
3                                   0.911504   
4                                   1.000000   

   INSTALL_INSTPAY_DIFF_DAYS_INSTALL_DAYS_PAY_MAX  \
0                                            36.0   
1                                            37.0   
2                                            38.0   
3                                            19.0   
4                                            18.0   

   INSTALL_INSTPAY_DIFF_DAYS_INSTALL_DAYS_PAY_MIN  \
0                                           -11.0   
1                                            -1.0   
2                            

In [10]:
# Merge for full set
# testing_df = pd.concat ([testing_df, features_testing_df], axis=1)
testing_df_merged = testing_df.merge(right=features_testing_df, on = 'SK_ID_CURR', how = 'left')

print (testing_df_merged.shape)
testing_df_merged.head()

(48744, 359)


SK_ID_CURR  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0      100001             0          135000.0    568800.0      20560.5   
1      100005             0           99000.0    222768.0      17370.0   
2      100013             0          202500.0    663264.0      69777.0   
3      100028             2          315000.0   1575000.0      49018.5   
4      100038             1          180000.0    625500.0      32067.0   

   REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  \
0                    0.018850      -19241          -2329            -5170.0   
1                    0.035792      -18064          -4469            -9118.0   
2                    0.019101      -20038          -4458            -2175.0   
3                    0.026392      -13976          -1866            -2000.0   
4                    0.010032      -13040          -2191            -4000.0   

   DAYS_ID_PUBLISH                       ...                         \
0             -812                       ...                          
1            -1623                       ...                          
2            -3503                       ...                          
3            -4208                       ...                          
4            -4262                       ...                          

   INSTALL_DAYS_ENTRY_PAYMENT_MEDIAN  INSTALL_DAYS_INSTALMENT_MAX  \
0                            -1715.0                      -1619.0   
1                             -585.0                       -466.0   
2                            -1383.0                        -14.0   
3                             -812.0                        -27.0   
4                             -634.0                       -457.0   

   INSTALL_DAYS_INSTALMENT_MIN  INSTALL_DAYS_INSTALMENT_MEDIAN  \
0                      -2916.0                         -1709.0   
1                       -706.0                          -586.0   
2                      -2705.0                         -1383.0   
3                      -1773.0                          -812.0   
4                       -787.0                          -622.0   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MAX  \
0                                       1.0   
1                                       1.0   
2                                       1.0   
3                                       1.0   
4                                       1.0   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MIN  \
0                                  1.000000   
1                                  1.000000   
2                                  0.000266   
3                                  0.030496   
4                                  1.000000   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MEAN  \
0                                   1.000000   
1                                   1.000000   
2                                   0.935484   
3                                   0.911504   
4                                   1.000000   

   INSTALL_INSTPAY_DIFF_DAYS_INSTALL_DAYS_PAY_MAX  \
0                                            36.0   
1                                            37.0   
2                                            38.0   
3                                            19.0   
4                                            18.0   

   INSTALL_INSTPAY_DIFF_DAYS_INSTALL_DAYS_PAY_MIN  \
0                                           -11.0   
1                                            -1.0   
2                                           -21.0   
3                                            -7.0   
4                                             9.0   

   INSTALL_INSTPAY_DIFF_DAYS_INSTALL_DAYS_PAY_MEAN  
0                                         7.285714  
1                                        23.555556  
2                                         5.180645  
3                                         3.000000  
4                                        12.250000  

[5 rows x 359 columns]

In [11]:
# Drop ID
training_df = training_df_merged.drop ('SK_ID_CURR', axis=1)
testing_df = testing_df_merged.drop ('SK_ID_CURR', axis=1)

In [12]:
training_df.head()

CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0             0          202500.0    406597.5      24700.5   
1             0          270000.0   1293502.5      35698.5   
2             0           67500.0    135000.0       6750.0   
3             0          135000.0    312682.5      29686.5   
4             0          121500.0    513000.0      21865.5   

   REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  \
0                    0.018801       -9461           -637            -3648.0   
1                    0.003541      -16765          -1188            -1186.0   
2                    0.010032      -19046           -225            -4260.0   
3                    0.008019      -19005          -3039            -9833.0   
4                    0.028663      -19932          -3038            -4311.0   

   DAYS_ID_PUBLISH  OWN_CAR_AGE  \
0            -2120          9.0   
1             -291          9.0   
2            -2531         26.0   
3            -2437          9.0   
4            -3458          9.0   

                        ...                         \
0                       ...                          
1                       ...                          
2                       ...                          
3                       ...                          
4                       ...                          

   INSTALL_DAYS_ENTRY_PAYMENT_MEDIAN  INSTALL_DAYS_INSTALMENT_MAX  \
0                             -312.0                        -25.0   
1                             -806.0                       -536.0   
2                             -763.0                       -724.0   
3                             -211.0                        -11.0   
4                             -852.5                        -14.0   

   INSTALL_DAYS_INSTALMENT_MIN  INSTALL_DAYS_INSTALMENT_MEDIAN  \
0                       -565.0                          -295.0   
1                      -2310.0                          -797.0   
2                       -784.0                          -754.0   
3                       -545.0                          -206.0   
4                      -2326.0                          -851.0   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MAX  \
0                                       1.0   
1                                       1.0   
2                                       1.0   
3                                       1.0   
4                                       1.0   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MIN  \
0                                   1.00000   
1                                   1.00000   
2                                   1.00000   
3                                   1.00000   
4                                   0.00005   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MEAN  \
0                                   1.000000   
1                                   1.000000   
2                                   1.000000   
3                                   1.000000   
4                                   0.954545   

   INSTALL_INSTPAY_DIFF_DAYS_INSTALL_DAYS_PAY_MAX  \
0                                            31.0   
1                                            14.0   
2                                            11.0   
3                                            77.0   
4                                            31.0   

   INSTALL_INSTPAY_DIFF_DAYS_INSTALL_DAYS_PAY_MIN  \
0                                            12.0   
1                                             1.0   
2                                             3.0   
3                                             1.0   
4                                           -12.0   

   INSTALL_INSTPAY_DIFF_DAYS_INSTALL_DAYS_PAY_MEAN  
0                                        20.421053  
1                                         7.160000  
2                                         7.666667  
3                                        19.375000  
4                                         3.63636

In [13]:
testing_df.head()

CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0             0          135000.0    568800.0      20560.5   
1             0           99000.0    222768.0      17370.0   
2             0          202500.0    663264.0      69777.0   
3             2          315000.0   1575000.0      49018.5   
4             1          180000.0    625500.0      32067.0   

   REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  \
0                    0.018850      -19241          -2329            -5170.0   
1                    0.035792      -18064          -4469            -9118.0   
2                    0.019101      -20038          -4458            -2175.0   
3                    0.026392      -13976          -1866            -2000.0   
4                    0.010032      -13040          -2191            -4000.0   

   DAYS_ID_PUBLISH  OWN_CAR_AGE  \
0             -812          9.0   
1            -1623          9.0   
2            -3503          5.0   
3            -4208          9.0   
4            -4262         16.0   

                        ...                         \
0                       ...                          
1                       ...                          
2                       ...                          
3                       ...                          
4                       ...                          

   INSTALL_DAYS_ENTRY_PAYMENT_MEDIAN  INSTALL_DAYS_INSTALMENT_MAX  \
0                            -1715.0                      -1619.0   
1                             -585.0                       -466.0   
2                            -1383.0                        -14.0   
3                             -812.0                        -27.0   
4                             -634.0                       -457.0   

   INSTALL_DAYS_INSTALMENT_MIN  INSTALL_DAYS_INSTALMENT_MEDIAN  \
0                      -2916.0                         -1709.0   
1                       -706.0                          -586.0   
2                      -2705.0                         -1383.0   
3                      -1773.0                          -812.0   
4                       -787.0                          -622.0   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MAX  \
0                                       1.0   
1                                       1.0   
2                                       1.0   
3                                       1.0   
4                                       1.0   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MIN  \
0                                  1.000000   
1                                  1.000000   
2                                  0.000266   
3                                  0.030496   
4                                  1.000000   

   INSTALL_INSTPAY_PAY_TO_INSTALL_RATIO_MEAN  \
0                                   1.000000   
1                                   1.000000   
2                                   0.935484   
3                                   0.911504   
4                                   1.000000   

   INSTALL_INSTPAY_DIFF_DAYS_INSTALL_DAYS_PAY_MAX  \
0                                            36.0   
1                                            37.0   
2                                            38.0   
3                                            19.0   
4                                            18.0   

   INSTALL_INSTPAY_DIFF_DAYS_INSTALL_DAYS_PAY_MIN  \
0                                           -11.0   
1                                            -1.0   
2                                           -21.0   
3                                            -7.0   
4                                             9.0   

   INSTALL_INSTPAY_DIFF_DAYS_INSTALL_DAYS_PAY_MEAN  
0                                         7.285714  
1                                        23.555556  
2                                         5.180645  
3                                         3.000000  
4                                        12.25000

In [14]:
# Load labels data
labels_df = load_credit_data ("y_labels.csv")
print (labels_df.shape)

y_train = labels_df['TARGET'].copy()

(307511, 1)


In [15]:
# ROC curve
from sklearn.metrics import roc_curve, auc

def plot_roc_curve(fpr, tpr, label=None):
    plt.plot(fpr, tpr, linewidth=2, label=label)
    plt.plot([0, 1], [0, 1], 'k--')
    plt.axis([0, 1, 0, 1])
    plt.xlabel('False Positive Rate', fontsize=16)
    plt.ylabel('True Positive Rate', fontsize=16)

### Random Forest

In [18]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

final_rf = RandomForestClassifier(random_state=123)

param_grid = {'n_estimators': [20, 250, 300], 'max_depth' : [1, 3, 5]}
grid_search_final_rf = GridSearchCV(estimator=final_rf, param_grid=param_grid , cv=5, scoring='roc_auc', verbose=100)

grid_search_final_rf.fit(training_df, y_train)

# Results of the grid search for best n_estimator
print(grid_search_final_rf.best_params_)
print ("------------")

# Results of the grid search in general
cvres = grid_search_final_rf.cv_results_
for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
    print(mean_score, params)

Fitting 5 folds for each of 9 candidates, totalling 45 fits
[CV] max_depth=1, n_estimators=20 ....................................
[CV]  max_depth=1, n_estimators=20, score=0.7154802300884572, total=   4.7s
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    5.9s remaining:    0.0s
[CV] max_depth=1, n_estimators=20 ....................................
[CV]  max_depth=1, n_estimators=20, score=0.7137236029392369, total=   4.0s
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:   11.1s remaining:    0.0s
[CV] max_depth=1, n_estimators=20 ....................................
[CV]  max_depth=1, n_estimators=20, score=0.7115496698763748, total=   4.6s
[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:   17.1s remaining:    0.0s
[CV] max_depth=1, n_estimators=20 ....................................
[CV]  max_depth=1, n_estimators=20, score=0.7145898342361188, total=   4.3s
[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:   22.7s remaining:    0.0s
[CV] max_depth=1, n_estimators=2

[CV]  max_depth=5, n_estimators=250, score=0.7462740982159621, total= 2.4min
[Parallel(n_jobs=1)]: Done  37 out of  37 | elapsed: 30.9min remaining:    0.0s
[CV] max_depth=5, n_estimators=250 ...................................
[CV]  max_depth=5, n_estimators=250, score=0.7391959148177718, total= 2.3min
[Parallel(n_jobs=1)]: Done  38 out of  38 | elapsed: 33.3min remaining:    0.0s
[CV] max_depth=5, n_estimators=250 ...................................
[CV]  max_depth=5, n_estimators=250, score=0.7505084363917072, total= 2.4min
[Parallel(n_jobs=1)]: Done  39 out of  39 | elapsed: 35.8min remaining:    0.0s
[CV] max_depth=5, n_estimators=250 ...................................
[CV]  max_depth=5, n_estimators=250, score=0.7469944670442894, total= 2.5min
[Parallel(n_jobs=1)]: Done  40 out of  40 | elapsed: 38.3min remaining:    0.0s
[CV] max_depth=5, n_estimators=300 ...................................
[CV]  max_depth=5, n_estimators=300, score=0.746853112400194, total= 3.0min
[Parallel(n_

In [43]:
### Forgot None!
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

final_rf = RandomForestClassifier(random_state=123)

param_grid = {'n_estimators': [20, 250, 300], 'max_depth' : [None]}
grid_search_final_rf = GridSearchCV(estimator=final_rf, param_grid=param_grid , cv=5, scoring='roc_auc', verbose=100)

grid_search_final_rf.fit(training_df, y_train)

# Results of the grid search for best n_estimator
print(grid_search_final_rf.best_params_)
print ("------------")

# Results of the grid search in general
cvres = grid_search_final_rf.cv_results_
for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
    print(mean_score, params) 

Fitting 5 folds for each of 3 candidates, totalling 15 fits
[CV] max_depth=None, n_estimators=20 .................................
[CV]  max_depth=None, n_estimators=20, score=0.693795355560664, total= 1.0min
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:  1.1min remaining:    0.0s
[CV] max_depth=None, n_estimators=20 .................................
[CV]  max_depth=None, n_estimators=20, score=0.6910975284639682, total= 1.1min
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:  2.3min remaining:    0.0s
[CV] max_depth=None, n_estimators=20 .................................
[CV]  max_depth=None, n_estimators=20, score=0.6871995847045846, total= 1.1min
[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:  3.4min remaining:    0.0s
[CV] max_depth=None, n_estimators=20 .................................
[CV]  max_depth=None, n_estimators=20, score=0.6938310056238336, total= 1.1min
[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:  4.5min remaining:    0.0s
[CV] max_depth=None, 

In [ ]:
## NONE was not any better, and was longer in comparion to max_depth = 5. As such, this parameter value
## can be ignored.

In [20]:
# Find BEST model
from sklearn.model_selection import cross_val_predict

forest_final_clf = grid_search_final_rf.best_estimator_
y_probas_final_forest = cross_val_predict(forest_final_clf, training_df, y_train, cv=3, method="predict_proba")
y_scores_final_forest = y_probas_final_forest[:, 1] 
fpr_final_forest, tpr_final_forest, thresholds_final_forest = roc_curve(y_train, y_scores_final_forest)

print ("AUC: ", auc(fpr_final_forest, tpr_final_forest))

AUC:  0.7467176950695793


In [21]:
predictions = forest_final_clf.predict_proba(testing_df)[:, 1]
submit = load_credit_data ("submit_labels.csv")
submit['TARGET'] = predictions
submit.to_csv('random_forest_features_final.csv', index = False)

### Logistic Regression

In [57]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV

log_rf = LogisticRegression(random_state=123)

param_grid = {'C': [0.01, 0.1, 1, 10, 100, 1000],'penalty': ['l1', 'l2']}

grid_search_log_rf = GridSearchCV(estimator=log_rf, param_grid=param_grid , cv=5, scoring='roc_auc', verbose=100)

grid_search_log_rf.fit(training_df, y_train)

# Results of the grid search for best n_estimator
print(grid_search_log_rf.best_params_)
print ("------------")

# Results of the grid search in general
cvres = grid_search_log_rf.cv_results_
for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
    print(mean_score, params)

Fitting 5 folds for each of 14 candidates, totalling 70 fits
[CV] C=0.001, penalty=l1 .............................................
[CV] ..... C=0.001, penalty=l1, score=0.730554573229131, total= 2.8min
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:  2.8min remaining:    0.0s
[CV] C=0.001, penalty=l1 .............................................
[CV] .... C=0.001, penalty=l1, score=0.7275172951734359, total= 3.8min
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:  6.6min remaining:    0.0s
[CV] C=0.001, penalty=l1 .............................................
[CV] .... C=0.001, penalty=l1, score=0.7271524403958223, total= 2.4min
[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:  9.1min remaining:    0.0s
[CV] C=0.001, penalty=l1 .............................................
[CV] .... C=0.001, penalty=l1, score=0.7303388893736781, total= 3.8min
[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed: 12.9min remaining:    0.0s
[CV] C=0.001, penalty=l1 ..........................

[CV] ........ C=1, penalty=l2, score=0.6477344132809604, total= 2.9min
[Parallel(n_jobs=1)]: Done  38 out of  38 | elapsed: 130.6min remaining:    0.0s
[CV] C=1, penalty=l2 .................................................
[CV] .......... C=1, penalty=l2, score=0.66134786368545, total= 5.4min
[Parallel(n_jobs=1)]: Done  39 out of  39 | elapsed: 135.9min remaining:    0.0s
[CV] C=1, penalty=l2 .................................................
[CV] ........ C=1, penalty=l2, score=0.6284092116880707, total= 2.2min
[Parallel(n_jobs=1)]: Done  40 out of  40 | elapsed: 138.2min remaining:    0.0s
[CV] C=10, penalty=l1 ................................................
[CV] ....... C=10, penalty=l1, score=0.7722763151890251, total= 7.7min
[Parallel(n_jobs=1)]: Done  41 out of  41 | elapsed: 145.9min remaining:    0.0s
[CV] C=10, penalty=l1 ................................................
[CV] ....... C=10, penalty=l1, score=0.7704469375730401, total= 8.0min
[Parallel(n_jobs=1)]: Done  42 out of

In [58]:
# Find BEST model
from sklearn.model_selection import cross_val_predict

log_final_clf = grid_search_log_rf.best_estimator_
y_probas_log_final = cross_val_predict(log_final_clf, training_df, y_train, cv=3, method="predict_proba")
y_scores_log_final = y_probas_log_final[:, 1] 
fpr_log_final, tpr_log_final, thresholds_log_final = roc_curve(y_train, y_scores_log_final)

print ("AUC: ", auc(fpr_log_final, tpr_log_final))

AUC:  0.7702487533647071


In [ ]:
predictions = log_final_clf.predict_proba(testing_df)[:, 1]
submit = load_credit_data ("submit_labels.csv")
submit['TARGET'] = predictions
submit.to_csv('log_features_final.csv', index = False)

### SGD

In [38]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(training_df)

X_tr = scaler.transform(training_df)
X_te = scaler.transform(testing_df)

In [39]:
from sklearn.linear_model import SGDClassifier
from sklearn.model_selection import GridSearchCV

model_sgd = SGDClassifier(random_state=123)

param_distributions = {'loss': ['log', 'modified_huber'], 'alpha': [1e-3, 1e-2, 1e-1, 1e0, 1e1, 1e2], 'penalty': ['l1', 'l2']}

grid_search_sgd_cv = GridSearchCV(model_sgd, param_distributions, cv=5, scoring='roc_auc', verbose=100) 
grid_search_sgd_cv.fit(X_tr, y_train)

# Results of the grid search for best n_estimator
print(grid_search_sgd_cv.best_params_)
print ("------------")

# Results of the grid search in general
cvres = grid_search_sgd_cv.cv_results_
for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
    print(mean_score, params)

Fitting 5 folds for each of 24 candidates, totalling 120 fits
[CV] alpha=0.001, loss=log, penalty=l1 ...............................
[CV]  alpha=0.001, loss=log, penalty=l1, score=0.7296241043774638, total=   8.1s
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    8.3s remaining:    0.0s
[CV] alpha=0.001, loss=log, penalty=l1 ...............................
[CV]  alpha=0.001, loss=log, penalty=l1, score=0.7213308875733616, total=   8.1s
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:   16.6s remaining:    0.0s
[CV] alpha=0.001, loss=log, penalty=l1 ...............................
[CV]  alpha=0.001, loss=log, penalty=l1, score=0.7173434124835253, total=   7.6s
[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:   24.5s remaining:    0.0s
[CV] alpha=0.001, loss=log, penalty=l1 ...............................
[CV]  alpha=0.001, loss=log, penalty=l1, score=0.7293571975012095, total=   7.8s
[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:   32.5s remaining:    0.0s
[CV] alpha

[CV]  alpha=0.01, loss=modified_huber, penalty=l2, score=0.760197572472802, total=   5.0s
[Parallel(n_jobs=1)]: Done  36 out of  36 | elapsed:  3.9min remaining:    0.0s
[CV] alpha=0.01, loss=modified_huber, penalty=l2 .....................
[CV]  alpha=0.01, loss=modified_huber, penalty=l2, score=0.7581525923162261, total=   5.0s
[Parallel(n_jobs=1)]: Done  37 out of  37 | elapsed:  4.0min remaining:    0.0s
[CV] alpha=0.01, loss=modified_huber, penalty=l2 .....................
[CV]  alpha=0.01, loss=modified_huber, penalty=l2, score=0.753338113776288, total=   4.8s
[Parallel(n_jobs=1)]: Done  38 out of  38 | elapsed:  4.1min remaining:    0.0s
[CV] alpha=0.01, loss=modified_huber, penalty=l2 .....................
[CV]  alpha=0.01, loss=modified_huber, penalty=l2, score=0.7615828691781145, total=   5.0s
[Parallel(n_jobs=1)]: Done  39 out of  39 | elapsed:  4.2min remaining:    0.0s
[CV] alpha=0.01, loss=modified_huber, penalty=l2 .....................
[CV]  alpha=0.01, loss=modified_hu

[CV]  alpha=1.0, loss=modified_huber, penalty=l1, score=0.5019850416202948, total=   6.3s
[Parallel(n_jobs=1)]: Done  72 out of  72 | elapsed:  7.8min remaining:    0.0s
[CV] alpha=1.0, loss=modified_huber, penalty=l1 ......................
[CV]  alpha=1.0, loss=modified_huber, penalty=l1, score=0.5, total=   6.8s
[Parallel(n_jobs=1)]: Done  73 out of  73 | elapsed:  7.9min remaining:    0.0s
[CV] alpha=1.0, loss=modified_huber, penalty=l1 ......................
[CV]  alpha=1.0, loss=modified_huber, penalty=l1, score=0.5, total=   8.4s
[Parallel(n_jobs=1)]: Done  74 out of  74 | elapsed:  8.0min remaining:    0.0s
[CV] alpha=1.0, loss=modified_huber, penalty=l1 ......................
[CV]  alpha=1.0, loss=modified_huber, penalty=l1, score=0.5, total=   7.8s
[Parallel(n_jobs=1)]: Done  75 out of  75 | elapsed:  8.1min remaining:    0.0s
[CV] alpha=1.0, loss=modified_huber, penalty=l2 ......................
[CV]  alpha=1.0, loss=modified_huber, penalty=l2, score=0.739485097796429, total=

[CV]  alpha=100.0, loss=modified_huber, penalty=l1, score=0.5, total=   8.2s
[CV] alpha=100.0, loss=modified_huber, penalty=l1 ....................
[CV]  alpha=100.0, loss=modified_huber, penalty=l1, score=0.5, total=   7.4s
[CV] alpha=100.0, loss=modified_huber, penalty=l1 ....................
[CV]  alpha=100.0, loss=modified_huber, penalty=l1, score=0.5, total=   7.6s
[CV] alpha=100.0, loss=modified_huber, penalty=l1 ....................
[CV]  alpha=100.0, loss=modified_huber, penalty=l1, score=0.5, total=   7.8s
[CV] alpha=100.0, loss=modified_huber, penalty=l2 ....................
[CV]  alpha=100.0, loss=modified_huber, penalty=l2, score=0.7351508776797162, total=   5.7s
[CV] alpha=100.0, loss=modified_huber, penalty=l2 ....................
[CV]  alpha=100.0, loss=modified_huber, penalty=l2, score=0.7360241573569776, total=   5.0s
[CV] alpha=100.0, loss=modified_huber, penalty=l2 ....................
[CV]  alpha=100.0, loss=modified_huber, penalty=l2, score=0.7298907446666525, tota

In [40]:
from sklearn.model_selection import cross_val_predict

sgd_final_clf = grid_search_sgd_cv.best_estimator_
y_probas_sgd_final = cross_val_predict(sgd_final_clf, X_tr, y_train, cv=3, method="predict_proba")
y_scores_sgd_final = y_probas_sgd_final[:, 1] 
fpr_sgd_final, tpr_sgd_final, thresholds_sgd_final = roc_curve(y_train, y_scores_sgd_final)

print ("AUC: ", auc(fpr_sgd_final, tpr_sgd_final))

AUC:  0.7497403082624194


In [41]:
predictions = grid_search_sgd_cv.predict_proba(X_te)[:, 1]
submit = load_credit_data ("submit_labels.csv")
submit['TARGET'] = predictions
submit.to_csv('sgd_features_final.csv', index = False)

### DecisionTree

In [46]:
from sklearn.tree import DecisionTreeClassifier

model_tree = DecisionTreeClassifier(random_state=123)

param_grid = {'max_depth' : [None, 1, 3, 5], 'criterion': ['gini', 'entropy']}
grid_search_tree = GridSearchCV(model_tree, param_grid, verbose=100, cv=5, scoring='roc_auc')

grid_search_tree.fit(training_df, y_train)

# Results of the grid search for best n_estimator
print(grid_search_tree.best_params_)
print ("------------")

# Results of the grid search in general
cvres = grid_search_tree.cv_results_
for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
    print(mean_score, params)

Fitting 5 folds for each of 8 candidates, totalling 40 fits
[CV] criterion=gini, max_depth=None ..................................
[CV]  criterion=gini, max_depth=None, score=0.5427444942785854, total= 2.2min
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:  2.2min remaining:    0.0s
[CV] criterion=gini, max_depth=None ..................................
[CV]  criterion=gini, max_depth=None, score=0.5478759723177477, total= 2.3min
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:  4.5min remaining:    0.0s
[CV] criterion=gini, max_depth=None ..................................
[CV]  criterion=gini, max_depth=None, score=0.5442868265060261, total= 2.2min
[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:  6.7min remaining:    0.0s
[CV] criterion=gini, max_depth=None ..................................
[CV]  criterion=gini, max_depth=None, score=0.5450702149601574, total= 2.2min
[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:  9.0min remaining:    0.0s
[CV] criterion=gini, max

[CV]  criterion=entropy, max_depth=5, score=0.7211989720711731, total=  19.2s
[Parallel(n_jobs=1)]: Done  37 out of  37 | elapsed: 22.3min remaining:    0.0s
[CV] criterion=entropy, max_depth=5 ..................................
[CV]  criterion=entropy, max_depth=5, score=0.7164000061915268, total=  19.1s
[Parallel(n_jobs=1)]: Done  38 out of  38 | elapsed: 22.7min remaining:    0.0s
[CV] criterion=entropy, max_depth=5 ..................................
[CV]  criterion=entropy, max_depth=5, score=0.7260463408708759, total=  19.3s
[Parallel(n_jobs=1)]: Done  39 out of  39 | elapsed: 23.0min remaining:    0.0s
[CV] criterion=entropy, max_depth=5 ..................................
[CV]  criterion=entropy, max_depth=5, score=0.7252957286783169, total=  19.2s
[Parallel(n_jobs=1)]: Done  40 out of  40 | elapsed: 23.3min remaining:    0.0s
[Parallel(n_jobs=1)]: Done  40 out of  40 | elapsed: 23.3min finished
{'criterion': 'entropy', 'max_depth': 5}
------------
0.545153889830594 {'criterion':

In [47]:
from sklearn.model_selection import cross_val_predict

tree_final_clf = grid_search_tree.best_estimator_
y_probas_tree_final = cross_val_predict(tree_final_clf, training_df, y_train, cv=3, method="predict_proba")
y_scores_tree_final = y_probas_tree_final[:, 1] 
fpr_tree_final, tpr_tree_final, thresholds_tree_final = roc_curve(y_train, y_scores_tree_final)

print ("AUC: ", auc(fpr_tree_final, tpr_tree_final))

AUC:  0.7238740728693391


In [48]:
predictions = tree_final_clf.predict_proba(testing_df)[:, 1]
submit = load_credit_data ("submit_labels.csv")
submit['TARGET'] = predictions
submit.to_csv('tree_features_final.csv', index = False)

### AdaBoost on DecisionTree

In [52]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeClassifier

model_ada = AdaBoostClassifier(DecisionTreeClassifier(max_depth=5, criterion='entropy'), random_state=123)

param_distributions = {'learning_rate': [0.2, 0.5, 0.7]}
grid_search_ada_cv = GridSearchCV(model_ada, param_distributions, cv=5, scoring='roc_auc', verbose=100) 

grid_search_ada_cv.fit(training_df, y_train)

# Results of the grid search for best n_estimator
print(grid_search_ada_cv.best_params_)
print ("------------")

# Results of the grid search in general
cvres = grid_search_ada_cv.cv_results_
for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
    print(mean_score, params)

Fitting 5 folds for each of 3 candidates, totalling 15 fits
[CV] learning_rate=0.2 ...............................................
[CV] ...... learning_rate=0.2, score=0.7530576072195488, total=15.6min
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed: 15.7min remaining:    0.0s
[CV] learning_rate=0.2 ...............................................
[CV] ...... learning_rate=0.2, score=0.7554646538718301, total=15.4min
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed: 31.1min remaining:    0.0s
[CV] learning_rate=0.2 ...............................................
[CV] ...... learning_rate=0.2, score=0.7472669405366369, total=16.4min
[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed: 47.6min remaining:    0.0s
[CV] learning_rate=0.2 ...............................................
[CV] ...... learning_rate=0.2, score=0.7565723137470367, total=16.8min
[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed: 64.5min remaining:    0.0s
[CV] learning_rate=0.2 .............................

In [53]:
# Find BEST model
from sklearn.model_selection import cross_val_predict

ada_final_clf = grid_search_ada_cv.best_estimator_
y_probas_ada_final = cross_val_predict(ada_final_clf, training_df, y_train, cv=3, method="predict_proba")
y_scores_ada_final = y_probas_ada_final[:, 1] 
fpr_ada_final, tpr_ada_final, thresholds_ada_final = roc_curve(y_train, y_scores_ada_final)

print ("AUC: ", auc(fpr_ada_final, tpr_ada_final))

AUC:  0.7458762836569657


In [54]:
predictions = ada_final_clf.predict_proba(testing_df)[:, 1]
submit = load_credit_data ("submit_labels.csv")
submit['TARGET'] = predictions
submit.to_csv('ada_features_final_dt.csv', index = False)

### Ensembles of 'Best'

In [27]:
from sklearn.ensemble import VotingClassifier
from sklearn.ensemble import RandomForestClassifier

# To make this notebook's output stable across runs
np.random.seed(123)

b_clf1 = forest_final_clf
b_clf2 = log_final_clf
b_clf3 = tree_final_clf
# b_clf4 = SGDClassifier (alpha=0.1, loss='log', penalty='l2')

b_eclf = VotingClassifier(estimators=[('rfor', b_clf1), ('logreg', b_clf2), ('dt', b_clf3)], weights=[1,1,1], voting='soft')
b_eclf = b_eclf.fit(training_df, y_train) 

NameError: name 'log_final_clf' is not defined

In [55]:
from sklearn.ensemble import VotingClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB

# To make this notebook's output stable across runs
np.random.seed(123)

b_clf1 = RandomForestClassifier(n_estimators=300, max_depth=5)
b_clf2 = LogisticRegression(C=1, penalty='l1')
b_clf3 = DecisionTreeClassifier (max_depth=5, criterion='entropy')
# b_clf4 = SGDClassifier (alpha=0.1, loss='log', penalty='l2')
# b_clf4 = GaussianNB() # NEW
b_eclf = VotingClassifier(estimators=[('rfor', b_clf1), ('logreg', b_clf2), ('dt', b_clf3)], weights=[1,1,1], voting='soft')


b_eclf = b_eclf.fit(training_df, y_train) 

In [56]:
from sklearn.model_selection import cross_val_predict

y_probas_bvote_final = cross_val_predict(b_eclf, training_df, y_train, cv=3, method="predict_proba")
y_scores_bvote_final = y_probas_bvote_final[:, 1] 
fpr_bvote_final, tpr_bvote_final, thresholds_bvote_final = roc_curve(y_train, y_scores_bvote_final)

print ("AUC: ", auc(fpr_bvote_final, tpr_bvote_final))

AUC:  0.7633885663309568


In [57]:
predictions = b_eclf.predict_proba(testing_df)[:, 1]
submit = load_credit_data ("submit_labels.csv")
submit['TARGET'] = predictions
submit.to_csv('best_voting_classifier_features_final.csv', index = False)

### Ensemble of Ensembles

In [14]:
from sklearn.metrics import roc_auc_score
from sklearn.ensemble import ExtraTreesClassifier, RandomForestClassifier, VotingClassifier, AdaBoostClassifier, GradientBoostingClassifier

# conda install py-xgboost
import xgboost as xgb

# To make this notebook's output stable across runs
np.random.seed(123)

clf1 = AdaBoostClassifier(n_estimators=300)
clf2 = ExtraTreesClassifier(n_estimators=300, max_depth=5)
clf3 = xgb.XGBClassifier(n_estimators=300, max_depth=5)
clf4 = GradientBoostingClassifier(n_estimators=300, max_depth=5)
eclf = VotingClassifier(estimators=[('ab', clf1), ('etc', clf2), ('xgb', clf3),('gbc', clf4)], weights=[1,1,1,1], voting='soft')
eclf = eclf.fit(training_df, y_train)

In [15]:
from sklearn.model_selection import cross_val_predict

y_probas_vote_final = cross_val_predict(eclf, training_df, y_train, cv=3, method="predict_proba")
y_scores_vote_final = y_probas_vote_final[:, 1] 
fpr_vote_final, tpr_vote_final, thresholds_vote_final = roc_curve(y_train, y_scores_vote_final)

print ("AUC: ", auc(fpr_vote_final, tpr_vote_final))

AUC:  0.784135630893227


In [16]:
predictions = eclf.predict_proba(testing_df)[:, 1]
submit = load_credit_data ("submit_labels.csv")
submit['TARGET'] = predictions
submit.to_csv('voting_classifier_ensem_features_final.csv', index = False)

### Plot  curves

In [7]:
plt.figure(figsize=(8, 6))
plot_roc_curve(fpr_final_forest, tpr_final_forest, "Random Forest")
plot_roc_curve(fpr_log_final, tpr_log_final, "Log Reg")
plot_roc_curve(fpr_sgd_final, tpr_sgd_final, "SGD")
plot_roc_curve(fpr_gbc_final, tpr_gbc_final, "GBC")
plot_roc_curve(fpr_ada_final, tpr_ada_final, "AdaBoost")
plot_roc_curve(fpr_bvote_final, tpr_bvote_final, "Ensemble - Best")
plot_roc_curve(fpr_vote_final, tpr_vote_final, "Ensemble - Ensemble")
plt.legend(loc="lower right", fontsize=16)
plt.title("Comparing Models")
plt.show()

NameError: name 'fpr_final_forest' is not defined

### Submission scores:

   1. Random Forest : 0.731
   2. Logistic Regression : 0.766 
   3. SGD : 0.739 
   4. AdaBoost Decision Tree : 0.739
       1. Decision Tree only (for comparison): ---          
   5. Ensemble of ensembles (voting=soft) : 0.780
   6. Ensemble of 'best' (voting=soft) : 0.758